File: src/amenominaka/commands.ts
// amenominaka — clean-room Kit command stack (omni.kit.commands).
//
// Mirrors `omni.kit.commands`: undoable Command objects + a CommandStack with
// undo / redo and a bounded history. Each Command captures its inverse so
// `undo()` reverses `do()`. Executing a new command clears the redo stack
// (standard linear undo history — no branching).
//
// ADR-2605261800 §D6 / D10.4 amenominaka.

/** Abstract undoable command. Subclasses implement `do()` + `undo()`. */
export abstract class Command {
  name = "";
  abstract do(): void;
  abstract undo(): void;
}

/** Set a key on a dict-like target; undo restores the prior value (or removes
 *  the key if it was absent). */
export class SetAttributeCommand<T extends Record<string, unknown>> extends Command {
  name = "SetAttribute";
  private hadKey = false;
  private prevValue: unknown = undefined;

  constructor(
    private readonly target: T,
    private readonly key: keyof T & string,
    private readonly value: unknown,
  ) {
    super();
  }

  do(): void {
    this.hadKey = this.key in this.target;
    this.prevValue = this.target[this.key];
    (this.target as Record<string, unknown>)[this.key] = this.value;
  }
  undo(): void {
    if (this.hadKey) (this.target as Record<string, unknown>)[this.key] = this.prevValue;
    else delete (this.target as Record<string, unknown>)[this.key];
  }
}

export class CommandStack {
  private undoStack: Command[] = [];
  private redoStack: Command[] = [];

  constructor(private readonly historySize = 1000) {}

  /** Execute a command and push it onto the undo stack. An exception in
   *  `do()` is rethrown and the command is NOT pushed. Clears the redo stack. */
  execute(cmd: Command): void {
    cmd.do();
    this.undoStack.push(cmd);
    this.redoStack = [];
    if (this.undoStack.length > this.historySize) this.undoStack.shift();
  }

  /** Undo the most recent command (no-op if empty). Returns it, or null. */
  undo(): Command | null {
    const cmd = this.undoStack.pop();
    if (!cmd) return null;
    cmd.undo();
    this.redoStack.push(cmd);
    return cmd;
  }

  /** Redo the most recently undone command (no-op if empty). */
  redo(): Command | null {
    const cmd = this.redoStack.pop();
    if (!cmd) return null;
    cmd.do();
    this.undoStack.push(cmd);
    return cmd;
  }

  canUndo(): boolean {
    return this.undoStack.length > 0;
  }
  canRedo(): boolean {
    return this.redoStack.length > 0;
  }
  history(): string[] {
    return this.undoStack.map((c) => c.name || c.constructor.name);
  }
  clear(): void {
    this.undoStack = [];
    this.redoStack = [];
  }
}

// ── global command stack (omni.kit.commands.execute/undo/redo) ───────────────

let _globalStack: CommandStack | null = null;

function globalStack(): CommandStack {
  if (_globalStack === null) _globalStack = new CommandStack();
  return _globalStack;
}

export function execute(cmd: Command): void {
  globalStack().execute(cmd);
}
export function undo(): Command | null {
  return globalStack().undo();
}
export function redo(): Command | null {
  return globalStack().redo();
}
/** Reset the global command stack (test helper). */
export function _resetStack(): void {
  _globalStack = null;
}

File: src/kotoba/lang/kami_nv_compat/amenominaka/commands.clj
(ns kotoba.lang.kami-nv-compat.amenominaka.commands
  "Clean-room Kit command stack (omni.kit.commands) — JVM port of
  src/amenominaka/commands.ts. Undoable commands + a CommandStack with
  undo/redo and a bounded history; executing a new command clears the redo
  stack (linear undo, no branching). Each command captures its inverse so
  undo! reverses do!.

  Wave-2 (sub) of the kami-nv-compat TS->CLJC port (ADR-2607020130): the
  amenominaka subdir is split across waves — this namespace (commands) is
  self-contained; the extension.toml parser + Application lifecycle land in
  follow-up waves.

  The TS models targets as plain mutable objects; here a target is an atom
  holding a map, and set-attribute-command mutates it via swap!.")

;; ── ICommand protocol (mirrors the abstract `Command` class) ───────────────

(defprotocol ICommand
  (command-do! [this] "Apply the command's effect.")
  (command-undo! [this] "Reverse command-do!.")
  (command-name [this] "Stable name for history listing."))

;; ── SetAttributeCommand ───────────────────────────────────────────────────

(defn set-attribute-command
  "A command that sets `key` to `value` on the atom-held map `target`; undo!
  restores the prior value (or removes the key if it was absent). Mirrors the
  TS SetAttributeCommand<T>."
  [target key value]
  (let [had-key (volatile! false)
        prev    (volatile! nil)]
    (reify ICommand
      (command-do! [_]
        (let [cur @target]
          (vreset! had-key (contains? cur key))
          (vreset! prev (get cur key))
          (swap! target assoc key value)))
      (command-undo! [_]
        (if @had-key
          (swap! target assoc key @prev)
          (swap! target dissoc key)))
      (command-name [_] "SetAttribute"))))

;; ── CommandStack ──────────────────────────────────────────────────────────

(defprotocol ICommandStack
  (execute! [this cmd] "Apply cmd, push it on the undo stack, clear redo; trim oldest past cap. Throws if do! throws (cmd not pushed).")
  (undo!    [this] "Undo the most recent command; returns it, or nil if empty.")
  (redo!    [this] "Redo the most recently undone command; returns it, or nil if empty.")
  (can-undo? [this])
  (can-redo? [this])
  (history [this] "Names of the undo stack, oldest-first.")
  (clear! [this]))

(defn command-stack
  "Returns a new ICommandStack with a bounded history (default 1000)."
  ([]
   (command-stack 1000))
  ([history-size]
   (let [state (atom {:undo [] :redo [] :history-size history-size})]
     (reify ICommandStack
       (execute! [_ cmd]
         (command-do! cmd)                       ; may throw → not pushed (TS parity)
         (swap! state
                (fn [{:keys [undo history-size]}]
                  (let [undo' (conj undo cmd)]
                    {:undo (if (> (count undo') history-size)
                             (subvec undo' 1)    ; drop oldest
                             undo')
                     :redo []
                     :history-size history-size}))))
       (undo! [_]
         (let [undo (:undo @state)]
           (if (empty? undo)
             nil
             (let [cmd (peek undo)]
               (command-undo! cmd)
               (swap! state (fn [s]
                              (-> s
                                  (update :undo pop)
                                  (update :redo conj cmd))))
               cmd))))
       (redo! [_]
         (let [redo (:redo @state)]
           (if (empty? redo)
             nil
             (let [cmd (peek redo)]
               (command-do! cmd)
               (swap! state (fn [s]
                              (-> s
                                  (update :redo pop)
                                  (update :undo conj cmd))))
               cmd))))
       (can-undo? [_] (some? (seq (:undo @state))))
       (can-redo? [_] (some? (seq (:redo @state))))
       (history  [_] (mapv command-name (:undo @state)))
       (clear!   [_] (swap! state assoc :undo [] :redo []))))))

;; ── global command stack (omni.kit.commands.execute/undo/redo) ────────────

(defonce ^:private global-stack (atom nil))

(defn- global-stack! []
  (or @global-stack (reset! global-stack (command-stack))))

(defn execute [cmd] (execute! (global-stack!) cmd))
(defn undo   []  (undo! (global-stack!)))
(defn redo   []  (redo! (global-stack!)))

(defn reset-stack!
  "Reset the global command stack (test helper; mirrors TS `_resetStack`)."
  []
  (reset! global-stack nil))
